### PR TITLE
Variable printf arguments: let compiler help with formatting

### PR DESCRIPTION
**Pull request by anisse**
This generates many warnings, it might not be suitable to merge as-is
without many fixes first.

Attribute supported by both gcc and clang.

 

**Your checklist for this pull request**
- [X] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [X] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

While playing with ESIL, I found that type and argument verification wasn't properly checked at compile time with r_strbuf_appendf ; this will give more compiler warnings in case of wrong arguments size, bad order, forgotten arguments, too many arguments, etc.

**Test plan**

Build, and you'll find many new warnings. There may be bugs there.

**Closing issues**

N/A

### CODE REVIEW

**Comment by XVilka** [comment]
It breaks MSVC build because there is no such thing for it: https://ci.appveyor.com/project/radareorg/radare2/builds/34634627/job/p8yamuogkauejlem

**Comment by anisse** [comment]
I can add compiler checking for this. It would be too bad to miss on these checks.

**Comment by trufae** [comment]
For portability reasons i would say to have this wrapped as a macro so we can disable it source-wide in one place. Also may be worth add this to other vararg-format functions. Cool!

**Comment by XVilka** [comment]
You might create something like `R_PACKED` macro to switch this automatically depending on the compiler: https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#packed-structures

**Comment by anisse** [comment]
I might have reached too far, looks like I broke some builds, let me check again.

**Comment by XVilka** [comment]
I recommend to fix these warnings in the same PR - we can then merge commit without squashing.

File: libr/include/r_types.h [hunk 1]
```diff
@@ -210,6 +210,13 @@
   #define FUNC_ATTR_ALWAYS_INLINE
 #endif
 
+/* printf format check attributes */
+#if defined(__clang__) || defined(__GNUC__)
+#define R_PRINTF_CHECK(fmt, dots) __attribute__ ((format (printf, fmt, dots)))
+#else
+#define R_PRINTF_CHECK(format, dots)
```
Comment (ret2libc) [inline libr/include/r_types.h:217]: use `fmt`, `dots` as in the previous definition.

File: libr/include/r_util/r_strbuf.h [hunk 2]
```diff
@@ -19,12 +19,14 @@ R_API bool r_strbuf_set(RStrBuf *sb, const char *s);
 R_API bool r_strbuf_slice(RStrBuf *sb, int from, int len);
 R_API bool r_strbuf_setbin(RStrBuf *sb, const ut8 *s, size_t len);
 R_API ut8* r_strbuf_getbin(RStrBuf *sb, int *len);
-R_API bool r_strbuf_setf(RStrBuf *sb, const char *fmt, ...);
+R_API bool r_strbuf_setf(RStrBuf *sb, const char *fmt, ...)
```
Comment (ret2libc) [inline libr/include/r_util/r_strbuf.h:22]: I know it becomes a bit longer, but I think putting it on the same line would be better. @trufae opinions?

**Review by ret2libc — COMMENTED** [review]
I really like this! Overall, this is a strong :+1: from me. Just have some comments, but I'd love to see this integrated as well! Thanks.

**Comment by anisse** [comment]
Thanks for your feedback @ret2libc . For this PR, I was thinking of splitting it in a first patch that adds the define, and maybe a few checks that have no warnings. And in followup PRs, adding the check and associated fixes for each of those functions. I'll fix your comments then.

**Comment by ret2libc** [comment]
@XVilka I'm ok with merging this with just the defines and fixing the reported warnings in separate PRs. The fact that this PR adds a bunch of warnings at compilation time is not @anisse 's fault, so waiting until everything is fixed just hides the real problems. WDYT?

**Comment by ret2libc** [comment]
> Thanks for your feedback @ret2libc . For this PR, I was thinking of splitting it in a first patch that adds the define, and maybe a few checks that have no warnings. And in followup PRs, adding the check and associated fixes for each of those functions. I'll fix your comments then.

Ok ignore what I said in the previous comment. @XVilka made me notice that if we merge this as-is everything will be red because of CodeQL check. So, I guess your plan is a better plan :)

File: libr/include/r_types.h (hunk 1 above)
Comment (anisse) [reply libr/include/r_types.h]: Fixed

File: libr/include/r_util/r_strbuf.h (hunk 2 above)
Comment (anisse) [reply libr/include/r_util/r_strbuf.h]: Fixed as well, it will be in a separate PR because it generates many warnings.

**Comment by anisse** [comment]
As discussed, I've tried to put the new warnings in separate PR #17529

**Comment by anisse** [comment]
It seems the macos tests timed-out. Should I rebase on master and retry ?

**Pull request merged**